### DISCUSSION TITLE
Question: dynamic components

### DISCUSSION THREAD

**Discussion by theSherwood** [luwes/sinuous#173] [Q&A]
I'm experimenting with some dynamic rendering of some components here:
https://codesandbox.io/s/dynamic-components-z69n5

I'm wrapping the dynamic component in a closure in order to make it reactive:
```js
const view = html`
    <h3>Dynamic Components</h3>
    <button onclick=${addToList}>
      Add to List
    </button>
    <button onclick=${switchComponent}>
      Switch Component
    </button>
    <hr />
    ${() =>
      html`
        <${list()} items=${items} />
      `}
  `;
```

Is there a more idiomatic way of dealing with dynamic components? And if not, would it be worth considering some alternative to a closure with a nested `html` call?

This feels related to #6 on Conditional Rendering, which faced a similar closure and nested `html` call issue. As RyanSolid pointed out, the template must be recreated repeatedly.

**Comment by luwes** [comment]
Thanks for your question, I can't think of a much better way to write this actually.

There's no getting around the closure and nested `html` call with template literals I'm afraid. (I've been thinking of adding the closures automatically but that will only work for JSX during the compilation step)

It looks a bit nicer if you make the closure a variable, but you lose locality...

```js
const List = () => html`
	<${list()} items=${items} />
`;

const List2 = itemsParam => () => html`
	<${list()} items=${itemsParam} />
`;

const view = html`
    <h3>Dynamic Components</h3>
    <button onclick=${addToList}>
      Add to List
    </button>
    <button onclick=${switchComponent}>
      Switch Component
    </button>
    <hr />
    ${List} // or ${List2(items)}
  `;
```

**Comment by theSherwood** [comment]
Interesting. Thanks for the quick reply and sorry to keep bombarding you with questions. But here are some more:

- Apart from the considerations of syntax, what are the performance consequences of the closure and having to recreate these templates (both for dynamic components and for conditional branching)?
- How would you anticipate this looking in JSX?
- Is the issue that this can't be expressed in template literals or that there exists no present convention for doing that?

Again, thank you for answering my questions. My understanding of some of this stuff is limited, but I'm interested in learning more, and potentially contributing if you are looking for contributors.

**Comment by luwes** [comment]
No worries. Glad to help. Valid questions 👍 

The closure is not gonna slow down much, it's negligible I think.

The recreating is easy to work around by hoisting it up and actually creating the DOM node or create a cache of it at the right time. The only thing that will happen with the switch is that the root node of the one component will be replaced by the other.

`memoization` would work well here I believe.

https://github.com/ryansolid/solid has a built in helper for this I believe.

I might have to take a look a that.

**Comment by luwes** [comment]
Related Solid issue: https://github.com/ryansolid/solid/issues/73

**Comment by theSherwood** [comment]
That was really informative. All of this inquiry is related to an application I'm planning in which dynamic components are going to be the norm. So I'm looking for a concise way to express dyamic components which is also performant. Since conditional branching exposes a similar issue in sinuous, I'm wanting to find a shorthand that abstracts away the root of both issues. But it looks like it will mean using some alternative to Tagged Templates (or maybe a modified Tagged Templates syntax) that incorporates memoization perhaps.

I'd like to avoid a compilation step, if possible, but it may be the case that it isn't possible.

Is there some idiomatic way of handling memoization in sinuous?

**Comment by luwes** [comment]
@theSherwood thanks for the feedback, valid requests.

I worked on your example to make this work with memoized components.
https://codesandbox.io/s/dynamic-components-zuwcz

There is no idiomatic way yet to handle this case in sinuous so I put together a working solution that might get a place in the core or a separate pkg. I'll have to think a little about it.

Hope this helps, let me know if you have any questions. All feedback is welcome.

**Comment by theSherwood** [comment]
@luwes Thanks for that example. That's definitely an improvement. With the design of this memoization functionality, is there a way the memo could be moved to the dynamic component rather than to the components the dynamic component is rendering? Some kind of observable memo?

So, rather than:
```
const component1 = memo(({ items }) => {
   ...
})
const component2 = memo(({ items }) => {
   ...
})
const list = o(component1);
```

...we can have:
```
const component1 = ({ items }) => {
   ...
}
const component2 = ({ items }) => {
   ...
}
const list = dynamic(component1);
```

...or maybe even:
```
const view = html`
    <h3>Dynamic Components</h3>
    <button onclick=${addToList}>
      Add to List
    </button>
    <button onclick=${switchComponent}>
      Switch Component
    </button>
    <hr />
    ${dynamic(list()({ items }))}
  `;

...

const component1 = ({ items }) => {
   ...
}
const component2 = ({ items }) => {
   ...
}
const list = o(component1);
```

`dynamic` may not be right for the api, but something like this feels a little cleaner than memoizing each of the components fed into the dynamic component. What do you think?

Could something like this also be used for conditional branching?

I also don't yet understand enough about the internals of the sinuous reactivity and rendering models to know whether this is practical.

**Comment by luwes** [comment]
took a few attempts but got it working!
https://codesandbox.io/s/dynamic-components-gl7ny

`memo` is the right tool for the job I think, it's possible to build a `dynamic` utility function like this one:

```js
const dynamic = memo(comp => {
  return memo(props => root(() => comp(props)));
});
```

Each function component that is passed will be used as the cache key so in this case the body of dynamic will only be ran 2 times ever.

they in turn return a memoized version of the component which is keyed on the props object. 

the `root` function is still needed because of the computations which have to live on even though the component is toggled on and off.

yup, this could all be used for conditional branching as well.

**Comment by theSherwood** [comment]
It seems to work until you try to use it twice in one component:
https://codesandbox.io/s/dynamic-components-5u5z8

**Comment by luwes** [comment]
yes, that's because of the first memo function. it will have the first execution cached of `dynamic` since the argument will be both `component1`. 
easy to get around it by providing a unique argument as 2nd argument.

https://codesandbox.io/s/dynamic-components-2-pc64t

edit: because it is cached it returns the same element, in DOM there cannot be the same element in two places.

**Comment by theSherwood** [comment]
Thanks. I've iterated on what you've done to try to make a slightly more streamlined api:
https://codesandbox.io/s/dynamic-components-4-0f4b1

So the closure has been cleaned away on the dynamic component and the rebuilding of the dom fragment is gone per your memo function. You can see the minimal function calls being executed in the `console.log` statements I left in the definition of `d`. So we now have a performant dynamic component syntax that looks like this:
```
${d(comp, props, children)}
```

Notice, that `comp` must be an observable, and the observable is passed into `d()`. Not `d(comp())`. 

So this feels like progress to me. However. I'd like to get here:
```
<${d(comp)} props=${props} >
    <p> ...assorted children... </p>
</${d(comp)}>
```

And I'm completely stymied as to how to do that. Is this possible without changing the way the tagged templates work?

**Comment by ryansolid** [comment]
FYI, `createMemo` in Solid isnt an actual Memo exactly. It's a pure computation. I just took Reacts naming convention. It actually is a computed with it's own reactive lifecycle internally. That being said from the outside pure computeds are essentially memos. Their arguments are just auto tracked dependencies rather than execution arguments.

**Comment by luwes** [comment]
thanks for checking in @ryansolid, ah yes I misremembered that.
I'll have to look more into `createMemo`.

I was wondering if Solid has a way to handle this issue of dynamic components?
I guess it is very similar to conditional branches.

The code of `sinuous/memo` could be used for Solid as well I think, it's pretty much a memoize function with one speciality to handle document fragments.

https://github.com/luwes/sinuous/blob/b95ae7cd522ca88f095fbc170c0146a0b129d5e8/packages/sinuous/memo/src/memo.js#L26-L32

---

@theSherwood this is the best I can get it
https://codesandbox.io/s/dynamic-components-4-s4ton

```js

html`<${dynamic(list)} items="${items}">
	I am a child
<//>`

const memoComp = memo(comp => {
  return memo((props, ...children) => {
    return root(() => comp(props, ...children));
  });
});

const dynamic = comp => {
  return (props, ...children) => {
    return () => memoComp(comp())(props, ...children);
  };
};
```

**Comment by theSherwood** [comment]
That looks great! There seems to be some kind of bug if the child isn't a text node, though. For example:
```
<${dynamic(list)} items="${items}">
   <p>I am a child.</p>
<//>
```
...causes the dynamic aspect of the component not to function. Did you notice that too? I haven't been able to figure out the cause.

**Comment by luwes** [comment]
@theSherwood that was a tricky one. The solution for that is here:
https://codesandbox.io/s/blazing-worker-0313z

The problem was that we're memoizing the components including the children.

So when the swap in the DOM happens the children are pulled out of the one component and put in the other. It's similar behavior as you would do an appendChild with a child that is already in the DOM. This was causing issues with clearing and adding of the component.

The solution is to create new children per component even though in this case they are exactly the same. They have to be new instances.

Shame it needs another closure and html tag, will check if this could be made better.

```js
  const view = html`
    <h3>Dynamic Components</h3>
    <button onclick=${addToList}>
      Add to List
    </button>
    <button onclick=${switchComponent}>
      Switch Component
    </button>
    <hr />
    <${dynamic(list)} items="${items}">
      ${() => html`
        <p>I am a child</p>
        <p>I am a child too</p>
      `}
    <//>
  `;
```

**Comment by ryansolid** [comment]
Hmm.. What I was doing was passing the Component as a prop to a "DynamicComponent" which made this problem very trivial as wrapping the results in a Computed will just dynamically insert. It was like 2 liner so I didn't even bother making a helper for it. But in JSX there is no way to change the Component dynamically.. sure you can assign it to a variable, but it can never change after that point since I'd have to re-evaluate the whole tag and trying to detect where that was the intention at compile time doesn't seem possible. So I basically treated it like a HOC with a normal conditional / ternary which is a lot easier.  The switching Component is just a syntax nicety but you can do it manually via abstraction. Structurally there is no benefit.

Your case is more interesting since you want to swap it out without redoing unnecessary work. But that's hard because as soon as you are calling the function again you are doing it. It's a completely different internal map. There is no diffing happening here. When I was talking about inefficient re-renders on conditionals I mean in cases where the conditionals re-run unnecessarily like `count() > 5`. Every increment to the count re-renders.

This isn't one of those cases. You are swapping Component1 with Component2 only when necessary. If it was a matter of preserving children that would doable as their render cycle can happen independently, however changing the Component and changing how things are mapped isn't going to happen. If you want to be able to change the mapping you have to do it at a finer grain. You can't rerun the Component if you expect it not to re-render the children. It would be better to in this case pass down a multiplier rather than make it a different Component. 

================
EDIT what @luwes just said above. Looks like we were typing responses at the same time.

**Comment by theSherwood** [comment]
@luwes @ryansolid 
Thank you both for your thorough responses! I really appreciate the time you both have taken on this.

Seems like the upshot is that there's really no clean way to do this even using a JSX compiler. I would need to make my own compiler to make it clean?

```
<${dynamic(list)} items="${items}">
   ${() => html`
       <p>I am a child</p>
       <p>I am a child too</p>
   `}
<//>
```

...is really close! But the nested closure and `html` call break the convention with how other components are used. And the only other alternative (sans some compiler that is built to handle these cases) is a dynamic component like @ryansolid talks about.

I have to give this some serious thought. The use case I had in mind was for a quine with a reactive store holding the source code for all the components. Components that render other components must fetch them from the store. So if a rendered component's source is edited, the rendered component should update in real-time. Creating a sort of snappy interactive programming environment. But if I'm creating this much trouble on the render api side, maybe I just need to rethink the architecture of this whole thing.

**Comment by luwes** [comment]
@ryansolid @theSherwood thanks, great points. in this case of switching components one by one it's probably easier to not memoize and re-render the component on each switch. 

I made it work without the extra `${() => html`` ` for the children locally in my dev environment but this will make the children variable a function. I'm a bit on the fence if this will be a right change to make. this looks more like an edge case with the memoization etc.

**Comment by ryansolid** [comment]
Yeah I was going to write another response earlier. But got caught up with work. I don't think the memoization makes sense. The original example functionally makes the most sense, maybe just not the syntax.  Having the component swap release the previous results and run the new component makes a ton of sense.

As for the idealized use case, it sounds like Hot Module Reloading, which is one of the few things I haven't managed to figure out for fine-grained reactive libraries. The reason is the reactive graph is like a single organism. You can amputate limbs and have them grow back but hot swapping in the middle requires the reactive library to have the ability to orphan and accept nodes. I've never seen it, but I've thought about it. Technically roots can allow for this separation but it is a heavy cost to not opt into. At that point component-based reactivity (like Svelte looks a bit more attractive as an approach).

The other piece of the puzzle is not throwing away descendants which is harder for any reactive approach. Without diffing how do you know what corresponds after a replacement. Now I suppose it would be possible to use similar techniques we use for hydration to repurpose existing nodes, but it is more than the DOM tree, as you need to reconstruct the graph.

I've done the first part with solid-element (which each web component has separate roots) but I still wipe out all downstream components and lose local component state. But haven't found the time to take this sort of idea further especially as I became more disillusioned with Web Components.

**Comment by theSherwood** [comment]
Yes. A kind of fine-grained Hot Module Reloading. I had thought that keeping child nodes could allow the state of the application to go relatively undisturbed even while components up the component tree were being edited and swapped out. I hadn't realized that HMR was an unsolved problem with fine-grained reactive libraries. So there is no known way to avoid recreating all downstream nodes.

I guess that leaves the syntax. So is the closest we can get syntactically to do something similar to what Svelte does with `svelte:component`?
```
<${Component} is=${dynamicComponent} props=${props}>
  <p>Children</p>
<//>
```
...or we can have the closure:
```
${() => html`
   <${dynamicComponent()} props=${props}>
        <p>Children</p>
   <//>
`}
```
...both of which introduce some friction. With compilation it might be possible to add in the closures, but barring that, do we have any other syntax option?

**Comment by ryansolid** [comment]
Yeah I mean the first "Svelte-like" version was what I was thinking. Using a HOC gives some power too. Technically it would be possible to preserve the passed in props.children (or is it just children in Sinuous?). It's only the internal children of the Component that are hard to preserve. If think about it the children outside being passed in are bound in a closure to the parent context so they can be made independent of the Component you are injecting. In short your first option there gives you the ability to intercept the incoming props and children to handle them in the smartest way possible for your scenario.

Other syntax options.. Hmm... Sinuous would not be able to tell if the component is dynamic or not since they are just functions. Even a compiler would have a hard time there without indication. I suppose if you forced a `<${dynamicComponent()}>` syntax and said because what is being passed in is a function execution rather than a var we know it could be reactive. But as I said limited to only Tagged Template literals as you can't do that with JSX.

The 
```
<${wrap(dynamicComponent)} props=${props}>
   <p>Children</p>
<//>
```
syntax could still work I think if desired and if it acted like the first example. It is still essentially creating a HOC if I understand it properly. Essentially it would return a component that dynamic executes the passed in component. I haven't tried this but wouldn't just removing the memo work (maybe with better null checking)?

```js
const wrap = comp => {
  // this is a HOC effectively, returning new factory Component
  return (props, ...children) => {
    // This is only runs once, so any preprocessing of comp can be done here,
    // as well as any caching of props/children under certain scenarios
    // as this is outside of any individual dynamic component's execution.
    // Maybe you want to fully resolve the children here in their own reactive closure
    // and cache them so you can pass the same to every component that gets swapped.

    // return the actual dynamic insert which is essentially a Child component
    return () => comp()(props, ...children);
  };
};
```

Aside:

A real challenge from an HMR standpoint is that modules aren't necessarily the natural boundaries. I could write a JS module with Components (that contain other Components) in it that contain zero owned reactive nodes. It's because the Components are just functions that don't own any state themselves. It's the parent reactive context that does which might not even be in the same file module. Now one could impose these boundaries for this purpose but it isn't something you'd do naturally with the library as it is unnecessary bloat. It's just a side effect of the efficiency of this approach that it isn't naturally setup with reactive boundaries to traverse at every arbitrary code modularization. Components have almost 0 cost here compared to other less granular reactive approaches like Svelte or Vue. A lot of these ideas we take for granted in modern dev came from leveraging the VDOM in ways that it excels in. We need to figure out if there are similar opportunities here, ... I mean beyond the raw performance, simpler abstraction over the DOM, and trivial effort for scheduling.

**Comment by theSherwood** [comment]
> Technically it would be possible to preserve the passed in props.children (or is it just children in Sinuous?)

I think it's just children in Sinuous.

> ...syntax could still work I think if desired and if it acted like the first example.

Which is the first example you are referring to?
```
${() =>
   html`
      <${list()} items=${items} />
`}
```
or
```
const List = () => html`
	<${list()} items=${items} />
`;

const List2 = itemsParam => () => html`
	<${list()} items=${itemsParam} />
`;

const view = html`
    <h3>Dynamic Components</h3>
    <button onclick=${addToList}>
      Add to List
    </button>
    <button onclick=${switchComponent}>
      Switch Component
    </button>
    <hr />
    ${List} // or ${List2(items)}
  `;
```
?

I'm trying both the simple wrap and the HOC without memoization here:
https://codesandbox.io/s/quizzical-dawn-qxnmg
I also wrapped the extra closure from the last example we had (https://codesandbox.io/s/blazing-worker-0313z).

These seem to be rendering correctly, but I can't tell how much rebuilding of the template is occurring. So I can't tell if this is progress or not. I don't understand how these are working to well enough to be able to give any kind of comparison of their performance characteristics. These seem like they might be the options, though.

Maybe my import from the store can simply import wrapped components, so that I can completely abstract the whole thing away from the user. 

@ryansolid I'm really interested in the HMR issue. If you have any further thoughts on this, I'd be interested in learning more. 

@luwes @ryansolid Just wanted to thank you both again for your time and expertise on this. It's invaluable and I love the work you both are doing.

**Edit:** I linked to the wrong Codesandbox. I have corrected that. It should make a lot more sense now.

**Comment by theSherwood** [comment]
Here's what I've got:
https://codesandbox.io/s/sharp-tu-xjcyp

This seems to have the basic setup for HMR but at a component level, rather than fine-grained. No memoization is happening. So the components are getting built every time, but the render api stays clean if we move all the wrapping into the store. That leaves the problem in the store of identifying which functions are components (and will need to be wrapped) and which functions are not. 

If one could dictate a coding style, one could, for example, say that all default exports will be wrapped as components; making for one component per file. Other functions and or state could then also be exported. That's less than ideal, to be sure. But may be workable? It *feels* like a start to some kind of HMR system in any case.

**Comment by ryansolid** [comment]
I did a quick check. This is working as well as you can expect. I double checked although the component children are being re-rendered the children you pass in fact are the exact same `p` tags and are just being attached and removed from the DOM as the component re-renders. Ie.. these only are created once:
```
<p>Items: ${items}</p>
<p>Items length: ${() => items().length}</p>
```
This ends up being similar to what I did with the Web Components, it wipes out all downstream and because you are wrapping each component in its own context you can do it. But you are right I was so hung up on making those boundaries I wasn't thinking about the fact that it can be done in the background by writing a webpack loader or parcel plugin. I always use a compiler I forget the bundler can work for you too. Using the default export is customary and is needed for things like lazy components and other things. I've written this before, this isn't that hard at all. Basically you generate a second file for each component at bundling time (atleast from a source map perspective) and it imports the component and wraps it in the HOC and then registers it with HMR stuff `module.accept` etc.. and then your handler will basically inject the new Component instance into the observable and trigger whenever it refreshes. All you need to do besides writing the plugin is maybe generalize a `hot` function that can be imported to remove redundant code. From there it's just creating the templated string for that file. I used https://github.com/jshmrtn/vue-hot-loader/blob/master/lib/loader.js as a base previously.

@theSherwood  In any case thanks for waking me up on this again. I will probably be adding this to Solid(Dom Expressions) soon. Even if imperfect it makes things a little nicer. I'm sure the people using Create Solid App would appreciate it at least a little bit. 

@luwes I imagine the difference between implementations with our libraries would be minimal. Literally basically the only difference being what is being used as the observable/signal etc.. This mechanism is just JS no tagged templates or JSX I imagine so it should be an easy port for whoever gets there first.

**Comment by theSherwood** [comment]
Hey, that's great! I hope you guys don't mind if I cannibalize whatever implementation is made in order to get my quine project to work.

**Comment by theSherwood** [comment]
As a follow up to the HMR issue:

If you have a file that looks something like this:
```
let state = o(0)

let component = () => html`
   <p>${state}</p>
   <button onclick=${() => state(state()+1)}> + </button>
`
```
...is it possible to not reset `state` to `0` when the file is edited and reloaded? I know this is possible with some bundlers in the clojurescript ecosystem and it makes for an incredibly fluid development experience. But clojurescript atoms may be easier for a bundler to detect than sinuous observables. One option may be coopting the `o` or `observable` function during dev bundling to add a property to observables (or observables' prototype) that can be detected by the bundler.

**Comment by ryansolid** [comment]
React and most VDOM libraries have that capability too. That's always the goal, I just have never seen it done in a fine grained reactive library.

Well if you assume that all data needs to be derived downward, we only need to track observables created during the initial component execution in that context and we can avoid everything created within subcontexts (like computeds) atleast initially. While that is wasted work from a rendering perspective since those computeds might hold the actual DOM nodes, atleast assuming all state is a product of props and top level state you could recreate the same DOM from the same state, and presumably since the top level only executes once and is a product of that information it's order would be predictable. Something as simple as an array could track it. The bundler couldn't do it on its own there would have to be a hook into the reactive system. You'd have to add something to the Owning context, but it seems not that much of an overhead I suppose. Have to think of how to make it non-intrusive. There are some trickier edge cases though like computeds that are expecting to run once and write to an observable, but the observable already has the data from the previous state so they might execute again (like async requests) and trigger stuff again when they complete (or say trigger loading states in the middle). Basically it would be hard to ever account for all side effects. This is much trickier than just getting updated Component code.

On an aside beware of hoisting observables like your example. Unless you intend to share them between all instances of your Component (like if there are more than one instance on the page) you want it inside the component function.

**Comment by ryansolid** [comment]
Yeah that assumption I made for maintaining stated doesn't work. Or at least isn't that simple. Besides the difficulty of handling state changes without a VDOM (or a first pass to know it has in fact changed, so you could be assigning state to the wrong observables), you need to really know all the way down the tree. You couldn't just stop single level. So all hooks for serializing state would have to execute all the way down, and be applied all the way down from the insertion point. You'd need to patch the whole reactive system I think which goes back to what I was originally saying. This comes down to all dependencies being set dynamically at run time, with React Hooks it boils down to a simple equality check on variables. What we'd need to figure out is if we can diff a reactive graph while we are creating it. It might be enough if we could at least assign nodes to the module their code exists in rather than where they are executed. I suspect someone more familiar with how this works in a library like React post Hooks might hold the key but I still am missing something. So I am dropping working on state preservation at least for now.

That being said getting the simple case to work was trivial. I've made it work in Webpack but I want to get Parcel and Rollup too so. I'm still working on getting a universal version working and will let you know. It is really simple though. I'm spending more time setting up an environment to test rather than writing the plugin.

**Comment by theSherwood** [comment]
Okay. Interesting. I've been wondering a bit about that. On sinuous observables there is a little property to identify it as an observable. Is there an similar mechanism in solid?
https://github.com/luwes/sinuous/blob/master/packages/sinuous/observable/src/observable.js

```
function observable(value) {
  // Tiny indicator that this is an observable function.
  data.$o = true;
  data._observers = new Set();
  ...
}
```

What if that `$o` were instead assigned a uuid? Could that be used by the HMR to diff the state?

> That being said getting the simple case to work was trivial. I've made it work in Webpack...

Can you share a link to that?

**Comment by ryansolid** [comment]
As for the UUID I was thinking of that. But I mean it would essentially be a sequence number, although would probably have to make it into some sort of path to track nesting. What I was getting at is that you need to handle everything below the swap point.

Serialize all node ids, and values down the tree on the dispose step of HMR. Then as you recreate and hit the same ids, set the values instead of the passed-in values. However, all computations would need to run again which is why paths are essential to ensure not new ids assigned each execution. 

But how to do this. Since it's creation order in the context that matters not the dependency graph. And you'd need to give computations ids as well since they'd need to be able to give their part of the path on independent updates. You'd almost need to setup a graph beside the actual dependency graph. Or I guess on top, another collection on each context that you could traverse downwards to serialize. In so the computation itself that wraps the component swap could start as the root node. Then how do you read from this? I suppose we could wrap it in a global context while executing but would need to do a bit juggling as you entered and exited context. 

This, of course, goes completely south the second you edit the order of observables or computations or add new ones. A VDOM library has the ability to know it has changed to a certain extent and it can throw away the node and revert to the default behaviour of replacing that Component. If we ever failed we could not salvage the rest of the tree. We'd probably want to just restart without preserving the state since we'd never know if we already assigned incorrectly. I'm not clear we could detect failure in many cases. Like if you swapped the order of 2 observables it would probably completely break things but we'd never know unless we checked dependency resolution as well and realized somehow a computation now has different deps than before(id changed when you changed the order). But what is incorrect? The observable value assignment or did you edit that computation on purpose? Basically, to be safe we throw away the whole component and all its descendants. There are many cases we could not safely save state.  I do wonder if it could be recoverable.

This is complicated. But I do see one other interesting thing here. The id system and ability to serialize actually could be useful for Dev Tools. It would be an overhead for size and performance, but introducing a dev mode that could be skipped in production builds like React does might be the pattern. You could inspect a component and see its local state. This is all ideas but would definitely be a bit of an undertaking.

----------------------------------

A webpack loader would look like this, I've dropped your component in as an example (since my version was obviously written in Solid https://github.com/ryansolid/solid-hot-loader). But setting this as a Webpack loader should do the trick:
```js
const loaderUtils = require('loader-utils');

function loadHmr(file) {
  return `
    import { o } from "sinuous";
    import Comp from ${file};

    const obsv = o(Comp),
      Wrapped = (props, ...children) => {
        return () => obsv()(props, () => children);
      };

    export default Wrapped;

    if (module.hot) {
      module.hot.accept(${file}, () => obsv(Comp));
    }
  `;
}

module.exports = function load() {};

module.exports.pitch = function pitch(remainingRequest) {
  const file = loaderUtils.stringifyRequest(this, '!!' + remainingRequest);
  const isProduction = this.minimize || process.env.NODE_ENV === 'production';

  if (this.cacheable) {
    this.cacheable();
  }

  if (isProduction) {
    return `export * from ${file};`;
  }

  return loadHmr(file);
};
```
This approach relies on the bundler to wrap each component so an approach like this would not work in Parcel anyway as it requires transformation. This does mean you have to be specific with the rules to ensure you are only wrapping Component files. This is an ES6 version, but CJS is possible but need to change to requires.

**Comment by luwes** [comment]
this would be awesome but sounds pretty complicated 😄 

wanted to ping here that I fixed 2 issues that made our memoized version not work.

https://github.com/luwes/sinuous/blob/a0f9e7ce9a47e36619d26be9f12b141c8d8d3e7f/packages/sinuous/h/src/insert.js#L31-L39

with 0.21.4 this example works like it should with multiple child elements, each component only gets created once. the children of the component are still the same in both components, so it's not possible to have multiple components in the dome at the same time like you would expect. that requires a `() => children`
https://codesandbox.io/s/dynamic-components-4-u9tu3

edit: actually this does work. unexpected in a good way 
https://codesandbox.io/s/dynamic-components-4-l05bu

**Comment by theSherwood** [comment]
@ryansolid There's a ton to unpack here. 

> This, of course, goes completely south the second you edit the order of observables or computations or add new ones.

Can that be detected? As in, preserve state in the way you've talked about (with a path) in all instances unless the order of observables/computations has changed? Recreate the subtree if the order of observables has changed? Because that would still be pretty great!

Thanks for the sample webpack plugin. I have no experience with writing anything for bundlers. So that is astonishingly minimal.

> This approach relies on the bundler to wrap each component so an approach like this would not work in Parcel anyway as it requires transformation.

I didn't realize that. There's no way to transform code on its way to Parcel? I'm guessing Rollup is more like Webpack in that regard?

> This does mean you have to be specific with the rules to ensure you are only wrapping Component files.

I've been meaning to ask about that. How are you handling that for Solid? Does that mean that to use the bundler, only components can ever be default exports from any javascript file?

Once again, thanks for this. This is awesome work.

@luwes Awesome! 

Could you explain a little more to me about the performance differences this creates. I'm struggling to create a model of it in my mind.

**Comment by luwes** [comment]
@theSherwood the difference is that the non-memoized version would still re-create the elements of the component with each swap. the memoized version only creates these once for each component. basically the functions that represent the components are only run once for the memoized version if the arguments stay the same. take a look in the console logs, it will be clear.

**Comment by ryansolid** [comment]
@theSherwood 
(Using JSX since it's a bit less verbose and I'm more familiar with syntax).

```jsx
function CompA() {
  const count = o(3);
  const name = o("John");
  return <div>
    <span>Clicked {count()} times</span>
    <span>Hello, {name()}<span>
  </div>
}

function CompB() {
  const name = o("John");
  const count = o(3);
  return <div>
    <span>Clicked {count()} times</span>
    <span>Hello, {name()}<span>
  </div>
}
```
I'm not sure how to tell the difference. You could look at the type of the value but it's unreliable. Using pure pathing it would just assign the wrong state in this scenario. If there is nothing else unique about the call I don't know how to differentiate.

It's possible to do this with either Parcel or Rollup I'm sure as some sort of transformation. It just isn't as simple. Like you could do this with Babel I imagine. Webpack loaders are built so that you return the transformed string makes it pretty easy.

All I did was put all the Components in a "components" folder for starters. Basically though all you need to do is make use of include/exclude paths. It's not ideal but I haven't figured out a different way to do this. It is so much easier to handle this at a modular level with fine-grained approaches.

**Comment by theSherwood** [comment]
@ryansolid So the problem is doing this statically, right? Or at runtime? Because, I'm wondering if you use the path technique you were talking about, can you also cache the function name (`fn.name`) or source (`fn.toString()`). If there is any kind of mismatch, don't try to preserve state. Is that a workable approach?

**Comment by ryansolid** [comment]
@theSherwood 
Which fn? The function I'd be talking about here is `o`, the generic reactive atom. I'm not sure how we achieve this even at runtime without having users write interesting code (like passing unique names in). If anything we could make a compiler automatically pass in extra debug information without the end-user being any wiser.

If you meant the fn being the component sure we could detect a change but I'm not clear how we could trust anything downstream at that point which pretty much gets us back to where we are right now.

**Comment by theSherwood** [comment]
@ryansolid Yeah. Sorry, I was unclear. I meant the observable returned from `o`. But of course that won't work because every observable is the same function. So it basically has to be done through static analysis if it's to be done at all.

**Comment by ryansolid** [comment]
That's what I got so far. Admittedly I hadn't gotten to that conclusion before this conversation so maybe there are other options. It is within my means with the compiler to pull that off, so that's something. It still seems very complicated, but I will get there at some point I figure as its basically the last feature that keeps Solid from React parity. While not compatible I believe Solid supports more React features currently than even Preact other than HMR. So it is definitely something I'm conscious of. And of course, if I can figure anything else I will share the wealth.

And by all means, keep ideas coming.

**Comment by DrSensor** [comment]
Hi, anyone knows how to do it using only `h` function?
```js
h(dynamic(menu))
```
Seems doesn't work 🤔

(me trying to avoid `htm` parsing process without using babel)

EDIT:
finally, this works
```
h([dynamic(menu)])
```
